File: push-client-python/bio_thread_io_manager.py

import socket
from bio_client import *
from socket_channel import *
import time
import threading

class BioThreadIoManager:
  def __init__(self, s: socket, client):
    self.client = client
    self.socketChannel = SocketChannel(s)
    self.running = False
  
  def send(self, msg):
    self.socketChannel.writeAndFlush(msg)
  
  def close(self):
    self.running = False
    self.socketChannel.close()

  def startThread(self):
    self.running = True

    self.heartbeatThread = threading.Thread(target=self.__heartbeat__)
    self.ioloopThread = threading.Thread(target=self.__ioloop__)
    self.heartbeatThread.start()
    self.ioloopThread.start()
  
  def __heartbeat__(self):
    while self.running:
      try:
        self.send('heartbeat-' + self.client.userId)
      except:
        print('发送心跳失败')
      time.sleep(10)
    print ('heartbeat stop!')
    self.close()
  
  def __ioloop__(self):
    while self.running:
      try:
        str = self.socketChannel.readLine()
        self.client.onMessage(str)
      except Exception as e:
        print('连接发生异常 5s　后重新连接')
        time.sleep(5)
        self.client.reconnect()
        break
    print ("ioloop thread stop!")
    self.close()

File: push-client-javascript/bio_thread_io_manager.js
const SocketChannel = require('./socket_channel').SocketChannel

class BioThreadIoManager{
  constructor(socket, client, callback){
    this.client = client
    this.socketChannel = new SocketChannel(socket, this.onMessage)
    this.running = false
    this.callback = callback
  }

  onMessage(msg){
    this.callback(msg)
  }

  send(msg){
    this.socketChannel.writeAndFlush(msg)
  }

  close(){
    this.running = false
    this.socketChannel.close()
  }

  startThread(){
    this.heartbeatTimer = setInterval(() => {
      if (!this.running) {
        console.log('hearbeat thread stop!')
        clearInterval(this.heartbeatTimer)
        return;
      }
      this.heartbeat()
    }, 10000)
  }

  heartbeat(){
    try {
      this.socketChannel.send('heartbeat-' + this.client.userId)
    }catch(e) {
      console.log('发送心跳失败')
    }
  }

}